### PR TITLE
Current stage of app

### PR DESCRIPTION
**Pull request by yegormk**
Doesn`t working correctly with selection of FE technology versions.
There is also no checking for email if it already exists.

### CODE REVIEW

File: src/app/app.component.html [hunk 1]
```diff
@@ -1 +1,108 @@
...
+                The value you inputted is 
+                {{ registerForm.controls.firstName.errors?.minlength.actualLength }} 
+                characters long.
+                It must be at least 
+                {{ registerForm.controls.firstName.errors?.minlength.requiredLength }} 
+                characters long.
+            </mat-error>
+            <mat-error *ngIf="registerForm.controls.firstName.hasError('maxlength')">
+                The value you passed is too long.
+            </mat-error>
```
Comment (yazheviks) [inline src/app/app.component.html]: for these, we can have single error (must be from n to n characters)

File: src/app/app.component.html [hunk 2]
```diff
@@ -1 +1,108 @@
...
+        </mat-form-field>
 
+        <mat-form-field class="example-full-width" appearance="fill">
+            <mat-label>Last Name</mat-label>
+            <input matInput placeholder="Enter Last Name" formControlName="lastName">
+            <mat-error *ngIf="registerForm.controls.lastName.hasError('required')">
```
Comment (yazheviks) [inline src/app/app.component.html:15]: you usually have check on required error, let's create function with field param that checks it

File: src/app/app.component.html [hunk 3]
```diff
@@ -1 +1,108 @@
...
+              <mat-option></mat-option>
+              <mat-option *ngFor="let option of listVersions | keyvalue" [value]="option.key">{{ option.key | titlecase }}</mat-option>
+            </mat-select>
+        </mat-form-field>
+
+        <mat-form-field appearance="fill">
+            <mat-label>Choose a version</mat-label>
+            <mat-select 
+            [disabled]="!registerForm.controls.framework.valid" 
+            [formControl]="registerForm.controls.frameworkVersion">
```
Comment (yazheviks) [inline src/app/app.component.html]: use formControlName

File: src/app/app.component.html [hunk 4]
```diff
@@ -1 +1,108 @@
...
+                <span 
+                *ngIf="(registerForm.controls.hobbies.value?.length || 0) > 1" 
+                class="example-additional-selection">
+                  (+{{(registerForm.controls.hobbies.value?.length || 0) - 1}} 
+                  {{registerForm.controls.hobbies.value?.length === 2 ? 'other' : 'others'}})
+                </span>
+              </mat-select-trigger>
+              <mat-option *ngFor="let topping of hobbiesList" [value]="topping">{{topping}}</mat-option>
+            </mat-select>
+            <mat-error *ngIf="registerForm.controls.hobbies.hasError('required')">
```
Comment (yazheviks) [inline src/app/app.component.html]: this should be not select but inputs to set custom values

File: src/app/app.component.ts [hunk 5]
```diff
@@ -1,10 +1,61 @@
...
+  ];
+  
+  minDate = new Date(new Date().getFullYear() - 115, 0, 1);
+  maxDate = new Date(new Date().getFullYear() + 1, 11, 31);
+
+  registerForm = new FormGroup({
+    firstName: new FormControl('', [
+      Validators.required,
+      Validators.minLength(3),
+      Validators.maxLength(15),
```
Comment (yazheviks) [inline src/app/app.component.ts]: I guess we don't need these validators on fields

File: src/app/app.component.html (hunk 2 above)
Comment (yegormk) [reply src/app/app.component.html:15]: I have some questions(doubts) and ideas about it. 
So I didn`t implement this in final version of app.

File: src/app/app.component.css [hunk 6]
```diff
@@ -0,0 +1,43 @@
+.main{
```
Comment (yazheviks) [inline src/app/app.component.css:1]: spaces before {

File: src/app/app.component.html [hunk 7]
```diff
@@ -1 +1,112 @@
...
+                (click)="gettingVersions(option.value)" 
+                [value]="option.key"
+              >{{ option.key | titlecase }}</mat-option>
+            </mat-select>
+        </mat-form-field>
+
+        <mat-form-field appearance="fill">
+            <mat-label>Choose a version</mat-label>
+            <mat-select 
+            [disabled]="!registerForm.controls.framework.valid"
```
Comment (yazheviks) [inline src/app/app.component.html:52]: spaces

File: src/app/email-async-validator.service.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  emailUniqueValidator(): AsyncValidatorFn {
+    return (control : AbstractControl): Observable<ValidationErrors | null> => {
+      return this.isEmailExists(control.value).pipe(
+        map((exists) => {
+          if (exists) {
+            return { emailExists : true };
+          } else {
+            this.emailLibrary.push(control.value);
```
Comment (yazheviks) [inline src/app/email-async-validator.service.ts:27]: better to do it in sendForm or smth like that

File: src/app/app.component.ts [hunk 9]
```diff
@@ -1,10 +1,80 @@
...
+      [new FormGroup({
+        name: new FormControl('', Validators.required),
+        duration: new FormControl('', Validators.required),
+      })
+    ])
+  })
+
+  gettingVersions(val: string[]){
+    this.frameworkVersionsArr = val;
+  }
```
Comment (yazheviks) [inline src/app/app.component.ts:62]: a bit too hard logic for such task

File: src/app/app.component.ts [hunk 10]
```diff
@@ -1,10 +1,80 @@
...
+
+  minDate = new Date(new Date().getFullYear() - 115, 0, 1);
+  maxDate = new Date(new Date().getFullYear() + 1, 11, 31);
+  frameworkVersionsArr = ['']
+
+  registerForm = new FormGroup({
+    firstName: new FormControl('', [
+      Validators.required,
+    ]),
```
Comment (yazheviks) [inline src/app/app.component.ts:29]: can be written in 1 row

File: src/app/app.component.ts [hunk 11]
```diff
@@ -1,10 +1,80 @@
...
+
+  listVersions = {       
+    angular: ['1.1.1', '1.2.1', '1.3.3'],
+    react: ['2.1.2', '3.2.4', '4.3.1'],
+    vue: ['3.3.1', '5.2.1', '5.1.3'],
+  }
+
+  minDate = new Date(new Date().getFullYear() - 115, 0, 1);
+  maxDate = new Date(new Date().getFullYear() + 1, 11, 31);
+  frameworkVersionsArr = ['']
```
Comment (yazheviks) [inline src/app/app.component.ts:24]: add type for such things, not initialize with ['']
not needed var actually

File: src/app/app.component.ts [hunk 12]
```diff
@@ -1,10 +1,80 @@
...
+  }
+
+  listVersions = {       
+    angular: ['1.1.1', '1.2.1', '1.3.3'],
+    react: ['2.1.2', '3.2.4', '4.3.1'],
+    vue: ['3.3.1', '5.2.1', '5.1.3'],
+  }
+
+  minDate = new Date(new Date().getFullYear() - 115, 0, 1);
+  maxDate = new Date(new Date().getFullYear() + 1, 11, 31);
```
Comment (yazheviks) [inline src/app/app.component.ts:23]: do we need it?

File: src/app/app.component.ts [hunk 13]
```diff
@@ -1,10 +1,80 @@
...
-  title = 'task';
+  constructor(public checkForUniqueEmail: EmailAsyncValidatorService){
+
+  }
+
+  listVersions = {       
+    angular: ['1.1.1', '1.2.1', '1.3.3'],
+    react: ['2.1.2', '3.2.4', '4.3.1'],
+    vue: ['3.3.1', '5.2.1', '5.1.3'],
+  }
```
Comment (yazheviks) [inline src/app/app.component.ts:20]: better to move it to constants folder

File: src/app/app.component.ts [hunk 14]
```diff
@@ -1,10 +1,80 @@
...
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.css']
 })
 export class AppComponent {
-  title = 'task';
+  constructor(public checkForUniqueEmail: EmailAsyncValidatorService){
+
+  }
```
Comment (yazheviks) [inline src/app/app.component.ts:14]: can be 1 row if nothing is inside

File: src/app/app.component.ts [hunk 15]
```diff
@@ -1,10 +1,80 @@
 import { Component } from '@angular/core';
+import { FormGroup, FormControl, Validators, FormArray } from '@angular/forms'
+
+import { EmailAsyncValidatorService } from './email-async-validator.service';
```
Comment (yazheviks) [inline src/app/app.component.ts:4]: great!

**Pull request merged**